File: dsa/li10_tree/bt/java/Bt_L114_FlattenBinaryTreeToLinkedList.java

import com.algos.topics.models.TreeNode;


public class Bt_L114_FlattenBinaryTreeToLinkedList {

	public static void main(String[] args) {
		Bt_L114_FlattenBinaryTreeToLinkedList obj = new Bt_L114_FlattenBinaryTreeToLinkedList();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		//obj.flatten(root);
		obj.flatten2(root);
	}

	/*
	 * Problem: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
	 * Idea: Idea is to traverse in reverse postorder and flatten
	 * Use a global prev variable to track the last explored node in postorder traversal and assign it to right of current node.Then update the current as next previous for next iteration.
	 *
	 * */
	// Time: O(n)
    // space: O(1)
	TreeNode prev = null;
    public void flatten(TreeNode root) {
        if(root == null)
			return;

        flatten(root.right);
        flatten(root.left);

        root.right = prev;
        root.left = null;
        prev = root;
    }

	// Idea:You can use Morris traversal to implement the same in interative mannar.
	// Instead of assigning prev.right to curr, assign it to curr.right and then update curr.right to left and left to null.
	// Time: O(n)
    // space: O(1)
	public void flatten2(TreeNode root) {
		TreeNode curr = root;
		while(curr != null){
			if(curr.left != null){
				TreeNode prev = curr.left;
				while(prev.right != null){
					prev = prev.right;
				}

				prev.right = curr.right;
				curr.right = curr.left;
				curr.left = null;
			}

			curr = curr.right;
		}
	}
}

File: dsa/patterns/10_trees/bt/python/bt_l114__flatten_binary_tree_to_linked_list.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
# Idea: Traverse in reverse postorder and flatten
# Use a global prev variable to track last explored node. Assign it to right of current node.
# Update current as next previous for next iteration.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def flatten(root):
    prev = None

    def helper(node):
        nonlocal prev
        if not node:
            return

        helper(node.right)
        helper(node.left)

        node.right = prev
        node.left = None
        prev = node

    helper(root)


# Idea: Use Morris-style traversal iteratively
# For each node with a left child, find rightmost node in left subtree
# Assign rightmost.right = current.right, then move left subtree to right
# Time: O(n)
# Space: O(1)
def flatten_iterative(root):
    curr = root
    while curr:
        if curr.left:
            prev = curr.left
            while prev.right:
                prev = prev.right
            prev.right = curr.right
            curr.right = curr.left
            curr.left = None
        curr = curr.right


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

flatten_iterative(root)

# Print flattened linked list
curr = root
while curr:
    print(curr.val, end=" -> " if curr.right else "")
    curr = curr.right
# Expected: 1 -> 3 -> 2 -> 6 -> 5 -> 4
